flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/container/FlinkContainersBuilder.java

```diff
@@ -0,0 +1,276 @@
+        return new FlinkContainers(jobManager, taskManagers, zookeeper, conf);
+    }
+
+    // --------------------------- Helper Functions -------------------------------------
+
+    private GenericContainer<?> buildJobManagerContainer() {
+        // Configure JobManager
+        final Configuration jobManagerConf = new Configuration();
+        jobManagerConf.addAll(this.conf);
+        // Build JobManager container
+        final ImageFromDockerfile jobManagerImage;
+        try {
+            jobManagerImage =
+                    new FlinkImageBuilder()
+                            .setConfiguration(jobManagerConf)
+                            .setLogProperties(logProperties)
+                            .asJobManager()
+                            .build();
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to build JobManager image", e);
+        }
+        final GenericContainer<?> jobManager = buildContainer(jobManagerImage);
+        // Setup network for JobManager
+        jobManager
+                .withNetworkAliases(FlinkContainers.JOB_MANAGER_HOSTNAME)
+                .withExposedPorts(jobManagerConf.get(RestOptions.PORT));
+        // Setup logger
+        if (this.logger != null) {
+            jobManager.withLogConsumer(new Slf4jLogConsumer(this.logger).withPrefix("JobManager"));
+        }
+        return jobManager;
+    }
+
+    private List<GenericContainer<?>> buildTaskManagerContainers() {
+        List<GenericContainer<?>> taskManagers = new ArrayList<>();
+        for (int i = 0; i < numTaskManagers; i++) {
+            // Configure TaskManager
+            final Configuration taskManagerConf = new Configuration();
+            taskManagerConf.addAll(this.conf);
+            final String taskManagerHostName = FlinkContainers.TASK_MANAGER_HOSTNAME_PREFIX + i;
+            taskManagerConf.set(TaskManagerOptions.HOST, taskManagerHostName);
+            // Build TaskManager container
+            final ImageFromDockerfile taskManagerImage;
+            try {
+                taskManagerImage =
+                        new FlinkImageBuilder()
+                                .setConfiguration(taskManagerConf)
+                                .setLogProperties(logProperties)
+                                .asTaskManager()
+                                .build();
+            } catch (Exception e) {
+                throw new RuntimeException("Failed to build TaskManager image", e);
+            }
+            final GenericContainer<?> taskManager = buildContainer(taskManagerImage);
+            // Setup network for TaskManager
+            taskManager.withNetworkAliases(taskManagerHostName);
+            // Setup logger
+            if (this.logger != null) {
+                taskManager.withLogConsumer(
+                        new Slf4jLogConsumer(this.logger).withPrefix("TaskManager-" + i));
+            }
+            taskManagers.add(taskManager);
```

Comment:
Nit: This almost equal to the container creation of the JM can you extract a common method?